presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class WeightedDoubleReservoirSample
+        implements Cloneable
+{
+    private int size;
+    private double[] elements;
+    private double[] weights;
+
+    public WeightedDoubleReservoirSample(int maxSamples)
+    {
+        this.elements = new double[maxSamples];
+        this.weights = new double[maxSamples];
+        assertValid();
+    }
+
+    protected WeightedDoubleReservoirSample(WeightedDoubleReservoirSample other)
+    {
+        this.elements = new double[other.elements.length];
+        this.weights = new double[other.weights.length];
+        assertValid();
+        System.arraycopy(other.elements, 0, elements, 0, elements.length);
+        System.arraycopy(other.weights, 0, weights, 0, weights.length);
+    }
+
+    public long getMaxSamples()
+    {
+        return elements.length;
+    }
+
+    public void add(double element, double weight)
+    {
+        double adjustedWeight = getAdjustedWeight(weight);
+
+        if (size < elements.length) {
+            elements[size] = element;
+            weights[size] = adjustedWeight;
+            ++size;
+            bubbleUp();
+            assertValid();
+            return;
+        }
+
+        if (adjustedWeight >= weights[0]) {
+            return;
+        }
+
+        elements[0] = element;
+        weights[0] = adjustedWeight;
+        bubbleDown();
+        assertValid();
+    }
+
+    protected double getAdjustedWeight(double weight)
+    {
+        checkArgument(weight >= 0, "Weight cannot be negative");
+        return Math.pow(
+                ThreadLocalRandom.current().nextDouble(),
+                1.0 / weight);
+    }
+
+    public void mergeWith(WeightedDoubleReservoirSample other)
+    {
+        IntStream.range(0, other.elements.length).forEach(i -> add(other.elements[i], other.weights[i]));
+    }
+
+    @Override
+    public WeightedDoubleReservoirSample clone()
+    {
+        return new WeightedDoubleReservoirSample(this);
+    }
+
+    public DoubleStream stream()
+    {
+        return Arrays.stream(elements);
+    }
+
+    private void assertValid()
```

Comment:
@mbasmanova 

What's the Presto way to mark code as debug-mode only?

Reply:
@atavory There is no such way. Why do you need this?

Reply:
To check the heap invariants. The check is a linear-time operation (which makes the heap meaningless). 

I will remove this from the code.